File: src/main/java/me/onixdev/user/data/RotationContainer.java
package me.onixdev.user.data;

import dev.onixac.api.user.data.IPlayerRotationData;
import me.onixdev.event.impl.PlayerRotationEvent;
import me.onixdev.user.OnixUser;
import me.onixdev.util.math.GraphUtil;
import me.onixdev.util.math.MathUtil;
import me.onixdev.util.net.MinecraftValues;

import java.util.*;


public class RotationContainer  implements IPlayerRotationData {
    private final OnixUser user;
    private float yaw;
    private float pitch;
    private float lastYaw;
    private float lastPitch;
    private float deltaYaw;
    private float deltaPitch;
    private float lastDeltaYaw;
    private float lastDeltaPitch;
    private float yawAccel;
    private float pitchAccel;
    private float lastYawAccel;
    private float lastPitchAccel;
    private float rawMouseDeltaX;
    private float rawMouseDeltaY;
    private float fuckedPredictedPitch;
    private float fuckedPredictedYaw;
    private float lastFuckedPredictedPitch;
    private float lastFuckedPredictedYaw;
    private boolean cinematic;
    private double finalSensitivity;
    private double mcpSensitivity;
    private final ArrayDeque<Integer> sensitivitySamples;
    private int sensitivity;
    private int mouseDeltaX;
    private int mouseDeltaY;

    private float lastJoltYaw;

    private float lastJoltPitch;

    private float joltYaw;

    private float joltPitch;
    private float smoothnessYaw = 0.0f;
    private float smoothnessPitch = 0.0f;
    private float consistency = 0.0f;
    private long lastSmooth = 0L;
    private long lastHighRate = 0L;
    private final List<Double> yawSamples = new ArrayList<>();
    private final List<Double> pitchSamples = new ArrayList<>();
    private boolean cinematicRotation = false;
    private int isTotallyNotCinematic = 0;

    public RotationContainer(final OnixUser user) {
        this.user = user;
        this.sensitivitySamples = new ArrayDeque<Integer>();
    }

    public void handle(double yaw, double pitch) {
        this.lastYaw = this.yaw;
        this.lastPitch = this.pitch;
        this.yaw = (float) yaw;
        this.pitch = (float) pitch;
        this.lastJoltYaw = this.joltYaw;
        this.lastJoltPitch = this.joltPitch;
        this.joltYaw = Math.abs(deltaYaw - lastDeltaYaw);
        this.joltPitch = Math.abs(deltaPitch - lastDeltaPitch);
        this.lastDeltaYaw = this.deltaYaw;
        this.lastDeltaPitch = this.deltaPitch;
        this.deltaYaw = (float) Math.abs(yaw - this.lastYaw);
        this.deltaPitch = (float) Math.abs(pitch - this.lastPitch);
        PlayerRotationEvent preRotation = new PlayerRotationEvent(false, yaw, pitch, deltaYaw, deltaPitch);
        user.handleEvent(preRotation);
        this.lastPitchAccel = this.pitchAccel;
        this.lastYawAccel = this.yawAccel;
        this.yawAccel = Math.abs(this.deltaYaw - this.lastDeltaYaw);
        this.pitchAccel = Math.abs(this.deltaPitch - this.lastDeltaPitch);
        final float f = (float) this.mcpSensitivity * 0.6f + 0.2f;
        final float gcd = f * f * f * 1.2f;
        this.rawMouseDeltaX = this.deltaYaw / gcd;
        this.rawMouseDeltaY = this.deltaPitch / gcd;
        this.mouseDeltaX = (int) (this.deltaYaw / gcd);
        this.mouseDeltaY = (int) (this.deltaPitch / gcd);
        this.processCinematic();
        final float expectedYaw = this.deltaYaw * 1.073742f + (float) (this.deltaYaw + 0.15);
        final float expectedPitch = this.deltaPitch * 1.073742f - (float) (this.deltaPitch - 0.15);
        final float pitchDiff = Math.abs(this.deltaPitch - expectedPitch);
        final float yawDiff = Math.abs(this.deltaYaw - expectedYaw);
        this.lastFuckedPredictedPitch = this.fuckedPredictedPitch;
        this.lastFuckedPredictedYaw = this.fuckedPredictedYaw;
        this.fuckedPredictedPitch = Math.abs(this.deltaPitch - pitchDiff);
        this.fuckedPredictedYaw = Math.abs(this.deltaYaw - yawDiff);
        this.smoothnessYaw = 10.0f - Math.abs(yawAccel - lastYawAccel);
        this.smoothnessPitch = 5.0f - Math.abs(pitchAccel - lastPitchAccel);
        this.consistency = 1.0f - Math.abs(1.0f - (deltaYaw / (deltaPitch + 0.0001f)));

        if (this.deltaPitch > 0.1 && this.deltaPitch < 25.0f) {
            this.processSensitivity();
        }
        PlayerRotationEvent postRotation = new PlayerRotationEvent(true, yaw, pitch, deltaYaw, deltaPitch);
        user.handleEvent(postRotation);
//        lastYaw = this.yaw;
//        lastPitch = this.pitch;
//        if (user.getMovementContainer().getSetbackLocation() != null) {
//            user.getMovementContainer().getSetbackLocation().setYaw((float) yaw);
//            user.getMovementContainer().getSetbackLocation().setPitch((float) pitch);
//        }
//        this.yaw = yaw;
//        this.pitch = pitch;
//        lastDeltaYaw = deltaYaw;
//        lastDeltaPitch = deltaPitch;
//        this.deltaYaw = Math.abs(yaw - this.lastYaw);
//        this.deltaPitch = Math.abs(pitch - this.lastPitch);
//        PlayerRotationEvent preRotation = new PlayerRotationEvent(false, yaw, pitch, deltaYaw, deltaPitch);
//        user.handleEvent(preRotation);
//        if (this.deltaPitch > 0.1 && this.deltaPitch < 25.0f) {
//            this.processSensitivity();
//        }
//        processCinematic();
//        //Тут будет просчет сенсы и тд
//        PlayerRotationEvent postRotation = new PlayerRotationEvent(true, yaw, pitch, deltaYaw, deltaPitch);
//        user.handleEvent(postRotation);
    }

    private void processSensitivity() {
        final float gcd = (float) MathUtil.getGcd(this.deltaPitch, this.lastDeltaPitch);
        final double sensitivityModifier = Math.cbrt(0.8333 * gcd);
        final double sensitivityStepTwo = 1.666 * sensitivityModifier - 0.3333;
        final double finalSensitivity = sensitivityStepTwo * 200.0;
        this.finalSensitivity = finalSensitivity;
        this.sensitivitySamples.add((int) finalSensitivity);
        if (this.sensitivitySamples.size() == 40) {
            this.sensitivity = MathUtil.getMode(this.sensitivitySamples);
            if (this.hasValidSensitivity()) {
                this.mcpSensitivity = MinecraftValues.SENSITIVITY_MCP_VALUES.get(this.sensitivity);
            }
            this.sensitivitySamples.clear();
        }
    }

    private void processCinematic() {
        long now = System.currentTimeMillis();

        double differenceYaw = Math.abs(deltaYaw - lastDeltaYaw);
        double differencePitch = Math.abs(deltaPitch - lastDeltaPitch);

        double joltYaw = Math.abs(differenceYaw - deltaYaw);
        double joltPitch = Math.abs(differencePitch - deltaPitch);

        boolean cinematic = (now - lastHighRate > 250L) || (now - lastSmooth < 9000L);

        if (joltYaw > 1.0 && joltPitch > 1.0) {
            lastHighRate = now;
        }

        yawSamples.add((double) deltaYaw);
        pitchSamples.add((double) deltaPitch);

        if (yawSamples.size() >= 20 && pitchSamples.size() >= 20) {
            Set<Double> shannonYaw = new HashSet<>();
            Set<Double> shannonPitch = new HashSet<>();
            List<Double> stackYaw = new ArrayList<>();
            List<Double> stackPitch = new ArrayList<>();

            for (Double yawSample : yawSamples) {
                stackYaw.add(yawSample);
                if (stackYaw.size() >= 10) {
                    shannonYaw.add(MathUtil.getSE(stackYaw));
                    stackYaw.clear();
                }
            }

            for (Double pitchSample : pitchSamples) {
                stackPitch.add(pitchSample);
                if (stackPitch.size() >= 10) {
                    shannonPitch.add(MathUtil.getSE(stackPitch));
                    stackPitch.clear();
                }
            }

            if (shannonYaw.size() != 1 || shannonPitch.size() != 1 ||
                    !shannonYaw.toArray()[0].equals(shannonPitch.toArray()[0])) {
                isTotallyNotCinematic = 20;
            }

            GraphUtil.GraphResult resultsYaw = GraphUtil.INSTANCE.getGraph(yawSamples);
            GraphUtil.GraphResult resultsPitch = GraphUtil.INSTANCE.getGraph(pitchSamples);

            int negativesYaw = resultsYaw.getNegatives();
            int negativesPitch = resultsPitch.getNegatives();
            int positivesYaw = resultsYaw.getPositives();
            int positivesPitch = resultsPitch.getPositives();

            if (positivesYaw > negativesYaw || positivesPitch > negativesPitch) {
                lastSmooth = now;
            }

            yawSamples.clear();
            pitchSamples.clear();
        }

        if (isTotallyNotCinematic > 0) {
            isTotallyNotCinematic--;
            cinematicRotation = false;
        } else {
            cinematicRotation = cinematic;
        }

    }

    public boolean hasValidSensitivity() {
        return this.sensitivity > 0 && this.sensitivity < 200;
    }

    public boolean isCinematicRotation() {
        return cinematicRotation;
    }

    public boolean hasValidSensitivityNormalaized() {
        return this.sensitivity > 0 && this.sensitivity < 269;
    }

    public float getYaw() {
        return this.yaw;
    }

    public float getPitch() {
        return this.pitch;
    }

    public float getLastYaw() {
        return this.lastYaw;
    }

    public float getLastPitch() {
        return this.lastPitch;
    }

    public float getDeltaYaw() {
        return this.deltaYaw;
    }

    public float getDeltaPitch() {
        return this.deltaPitch;
    }

    public float getLastDeltaYaw() {
        return this.lastDeltaYaw;
    }

    public float getLastDeltaPitch() {
        return this.lastDeltaPitch;
    }

    public float getYawAccel() {
        return this.yawAccel;
    }

    public float getPitchAccel() {
        return this.pitchAccel;
    }

    public float getLastYawAccel() {
        return this.lastYawAccel;
    }

    public float getLastPitchAccel() {
        return this.lastPitchAccel;
    }

    public float getRawMouseDeltaX() {
        return this.rawMouseDeltaX;
    }

    public float getRawMouseDeltaY() {
        return this.rawMouseDeltaY;
    }

    public float getFuckedPredictedPitch() {
        return this.fuckedPredictedPitch;
    }

    public float getFuckedPredictedYaw() {
        return this.fuckedPredictedYaw;
    }

    public float getLastFuckedPredictedPitch() {
        return this.lastFuckedPredictedPitch;
    }

    public float getLastFuckedPredictedYaw() {
        return this.lastFuckedPredictedYaw;
    }

    public boolean isCinematic() {
        return this.cinematic;
    }

    public double getFinalSensitivity() {
        return this.finalSensitivity;
    }

    public double getMcpSensitivity() {
        return this.mcpSensitivity;
    }

    public ArrayDeque<Integer> getSensitivitySamples() {
        return this.sensitivitySamples;
    }

    public int getSensitivity() {
        return this.sensitivity;
    }

    public int getMouseDeltaX() {
        return this.mouseDeltaX;
    }

    public int getMouseDeltaY() {
        return this.mouseDeltaY;
    }

    public float getSmoothnessYaw() {
        return smoothnessYaw;
    }

    public float getSmoothnessPitch() {
        return smoothnessPitch;
    }

    public float getConsistency() {
        return consistency;
    }

    public void setCinematic(boolean cinematic) {
        this.cinematic = cinematic;
    }
}

File: src/main/java/me/onixdev/user/data/RotationContainer.kt
package me.onixdev.user.data

import dev.onixac.api.user.data.IPlayerRotationData
import me.onixdev.event.impl.PlayerRotationEvent
import me.onixdev.user.OnixUser
import me.onixdev.util.math.GraphUtil
import me.onixdev.util.math.MathUtil
import me.onixdev.util.net.MinecraftValues
import me.onixdev.util.rotation.Rotation
import java.util.*
import kotlin.math.abs
import kotlin.math.cbrt

class RotationContainer(private val user: OnixUser) : IPlayerRotationData {
    private var yaw = 0f
    private var pitch = 0f
    private var lastYaw = 0f
    private var lastPitch = 0f
    private var deltaYaw = 0f
    private var deltaPitch = 0f
    private var lastDeltaYaw = 0f
    private var lastDeltaPitch = 0f
    private var yawAccel = 0f
    var deltaYawABS = 0f
    var deltaPitchABS = 0f
    var lastDeltaYawABS = 0f
    var lastDeltaPitchABS = 0f
    private var pitchAccel = 0f
    private var lastYawAccel = 0f
    private var lastPitchAccel = 0f
    private var rawMouseDeltaX = 0f
    private var rawMouseDeltaY = 0f
    private var fuckedPredictedPitch = 0f
    private var fuckedPredictedYaw = 0f
    private var lastFuckedPredictedPitch = 0f
    private var lastFuckedPredictedYaw = 0f
    private var cinematic = false
    private var finalSensitivity = 0.0
    private var mcpSensitivity = 0.0
    private val sensitivitySamples: ArrayDeque<Int?> = ArrayDeque<Int?>()
    private var sensitivity = 0
    private var mouseDeltaX = 0
    private var mouseDeltaY = 0

    private var lastJoltYaw = 0f

    private var lastJoltPitch = 0f

    private var joltYaw = 0f

    private var joltPitch = 0f
    private var smoothnessYaw = 0.0f
    private var smoothnessPitch = 0.0f
    private var consistency = 0.0f
    private var lastSmooth = 0L
    private var lastHighRate = 0L
    private val yawSamples: MutableList<Double?> = ArrayList<Double?>()
    private val pitchSamples: MutableList<Double?> = ArrayList<Double?>()
    private var cinematicRotation = false
    private var isTotallyNotCinematic = 0

    fun handle(yaw: Double, pitch: Double) {
        user.rotation = Rotation(yaw, pitch)
        this.lastYaw = this.yaw
        this.lastPitch = this.pitch
        this.yaw = yaw.toFloat()
        this.pitch = pitch.toFloat()
        this.lastJoltYaw = this.joltYaw
        this.lastJoltPitch = this.joltPitch
        this.joltYaw = abs(deltaYaw - lastDeltaYaw)
        this.joltPitch = abs(deltaPitch - lastDeltaPitch)
        this.lastDeltaYawABS = this.deltaYawABS
        this.lastDeltaPitchABS = this.deltaPitchABS
        this.deltaYawABS = abs(yaw - this.lastYaw).toFloat()
        this.deltaPitchABS = abs(pitch - this.lastPitch).toFloat()
        this.lastDeltaYaw = this.deltaYaw
        this.lastDeltaPitch = this.deltaPitch
        this.deltaYaw = abs(yaw - this.lastYaw).toFloat()
        this.deltaPitch = abs(pitch - this.lastPitch).toFloat()
        val preRotation = PlayerRotationEvent(false, yaw, pitch, deltaYaw.toDouble(), deltaPitch.toDouble())
        user.handleEvent(preRotation)
        this.lastPitchAccel = this.pitchAccel
        this.lastYawAccel = this.yawAccel
        this.yawAccel = abs(this.deltaYaw - this.lastDeltaYaw)
        this.pitchAccel = abs(this.deltaPitch - this.lastDeltaPitch)
        val f = this.mcpSensitivity.toFloat() * 0.6f + 0.2f
        val gcd = f * f * f * 1.2f
        this.rawMouseDeltaX = this.deltaYaw / gcd
        this.rawMouseDeltaY = this.deltaPitch / gcd
        this.mouseDeltaX = (this.deltaYaw / gcd).toInt()
        this.mouseDeltaY = (this.deltaPitch / gcd).toInt()
        this.processCinematic()
        val expectedYaw = this.deltaYaw * 1.073742f + (this.deltaYaw + 0.15).toFloat()
        val expectedPitch = this.deltaPitch * 1.073742f - (this.deltaPitch - 0.15).toFloat()
        val pitchDiff = abs(this.deltaPitch - expectedPitch)
        val yawDiff = abs(this.deltaYaw - expectedYaw)
        this.lastFuckedPredictedPitch = this.fuckedPredictedPitch
        this.lastFuckedPredictedYaw = this.fuckedPredictedYaw
        this.fuckedPredictedPitch = abs(this.deltaPitch - pitchDiff)
        this.fuckedPredictedYaw = abs(this.deltaYaw - yawDiff)
        this.smoothnessYaw = 10.0f - abs(yawAccel - lastYawAccel)
        this.smoothnessPitch = 5.0f - abs(pitchAccel - lastPitchAccel)
        this.consistency = 1.0f - abs(1.0f - (deltaYaw / (deltaPitch + 0.0001f)))

        if (this.deltaPitch > 0.1 && this.deltaPitch < 25.0f) {
            this.processSensitivity()
        }
        val postRotation = PlayerRotationEvent(true, yaw, pitch, deltaYaw.toDouble(), deltaPitch.toDouble())
        user.handleEvent(postRotation)
        //        lastYaw = this.yaw;
//        lastPitch = this.pitch;
//        if (user.getMovementContainer().getSetbackLocation() != null) {
//            user.getMovementContainer().getSetbackLocation().setYaw((float) yaw);
//            user.getMovementContainer().getSetbackLocation().setPitch((float) pitch);
//        }
//        this.yaw = yaw;
//        this.pitch = pitch;
//        lastDeltaYaw = deltaYaw;
//        lastDeltaPitch = deltaPitch;
//        this.deltaYaw = Math.abs(yaw - this.lastYaw);
//        this.deltaPitch = Math.abs(pitch - this.lastPitch);
//        PlayerRotationEvent preRotation = new PlayerRotationEvent(false, yaw, pitch, deltaYaw, deltaPitch);
//        user.handleEvent(preRotation);
//        if (this.deltaPitch > 0.1 && this.deltaPitch < 25.0f) {
//            this.processSensitivity();
//        }
//        processCinematic();
//        //Тут будет просчет сенсы и тд
//        PlayerRotationEvent postRotation = new PlayerRotationEvent(true, yaw, pitch, deltaYaw, deltaPitch);
//        user.handleEvent(postRotation);
    }

    private fun processSensitivity() {
        val gcd = MathUtil.getGcd(this.deltaPitch.toDouble(), this.lastDeltaPitch.toDouble()).toFloat()
        val sensitivityModifier = cbrt(0.8333 * gcd)
        val sensitivityStepTwo = 1.666 * sensitivityModifier - 0.3333
        val finalSensitivity = sensitivityStepTwo * 200.0
        this.finalSensitivity = finalSensitivity
        this.sensitivitySamples.add(finalSensitivity.toInt())
        if (this.sensitivitySamples.size == 40) {
            this.sensitivity = MathUtil.getMode(this.sensitivitySamples)
            if (this.hasValidSensitivity()) {
                this.mcpSensitivity = MinecraftValues.SENSITIVITY_MCP_VALUES[this.sensitivity]!!
            }
            this.sensitivitySamples.clear()
        }
    }
    fun hasTooLowSensitivity(): Boolean {
        return finalSensitivity > -5 && finalSensitivity < 40
    }

    @Suppress("UNCHECKED_CAST")
    private fun processCinematic() {
        val now = System.currentTimeMillis()

        val differenceYaw = abs(deltaYaw - lastDeltaYaw).toDouble()
        val differencePitch = abs(deltaPitch - lastDeltaPitch).toDouble()

        val joltYaw = abs(differenceYaw - deltaYaw)
        val joltPitch = abs(differencePitch - deltaPitch)

        val cinematic = (now - lastHighRate > 250L) || (now - lastSmooth < 9000L)

        if (joltYaw > 1.0 && joltPitch > 1.0) {
            lastHighRate = now
        }

        yawSamples.add(deltaYaw.toDouble())
        pitchSamples.add(deltaPitch.toDouble())

        if (yawSamples.size >= 20 && pitchSamples.size >= 20) {
            val shannonYaw: MutableSet<Double?> = HashSet<Double?>()
            val shannonPitch: MutableSet<Double?> = HashSet<Double?>()
            val stackYaw: MutableList<Double?> = ArrayList<Double?>()
            val stackPitch: MutableList<Double?> = ArrayList<Double?>()

            for (yawSample in yawSamples) {
                stackYaw.add(yawSample)
                if (stackYaw.size >= 10) {
                    shannonYaw.add(MathUtil.getSE(stackYaw))
                    stackYaw.clear()
                }
            }

            for (pitchSample in pitchSamples) {
                stackPitch.add(pitchSample)
                if (stackPitch.size >= 10) {
                    shannonPitch.add(MathUtil.getSE(stackPitch))
                    stackPitch.clear()
                }
            }

            if (shannonYaw.size != 1 || shannonPitch.size != 1 || (shannonYaw.toTypedArray()[0] != shannonPitch.toTypedArray()[0])) {
                isTotallyNotCinematic = 20
            }

            val resultsYaw = GraphUtil.getGraph(yawSamples as List<Double>)
            val resultsPitch = GraphUtil.getGraph(pitchSamples as List<Double>)

            val negativesYaw = resultsYaw.negatives
            val negativesPitch = resultsPitch.negatives
            val positivesYaw = resultsYaw.positives
            val positivesPitch = resultsPitch.positives

            if (positivesYaw > negativesYaw || positivesPitch > negativesPitch) {
                lastSmooth = now
            }

            yawSamples.clear()
            pitchSamples.clear()
        }

        if (isTotallyNotCinematic > 0) {
            isTotallyNotCinematic--
            cinematicRotation = false
        } else {
            cinematicRotation = cinematic
        }
    }

    override fun hasValidSensitivity(): Boolean {
        return this.sensitivity in 1..<200
    }

    override fun isCinematicRotation(): Boolean {
        return cinematicRotation
    }

    override fun hasValidSensitivityNormalaized(): Boolean {
        return this.sensitivity in 1..<269
    }

    override fun getYaw(): Float {
        return this.yaw
    }

    override fun getPitch(): Float {
        return this.pitch
    }

    override fun getLastYaw(): Float {
        return this.lastYaw
    }

    override fun getLastPitch(): Float {
        return this.lastPitch
    }

    override fun getDeltaYaw(): Float {
        return this.deltaYaw
    }

    override fun getDeltaPitch(): Float {
        return this.deltaPitch
    }

    override fun getLastDeltaYaw(): Float {
        return this.lastDeltaYaw
    }

    override fun getLastDeltaPitch(): Float {
        return this.lastDeltaPitch
    }

    override fun getYawAccel(): Float {
        return this.yawAccel
    }

    override fun getPitchAccel(): Float {
        return this.pitchAccel
    }

    override fun getLastYawAccel(): Float {
        return this.lastYawAccel
    }

    override fun getLastPitchAccel(): Float {
        return this.lastPitchAccel
    }

    override fun getRawMouseDeltaX(): Float {
        return this.rawMouseDeltaX
    }

    override fun getRawMouseDeltaY(): Float {
        return this.rawMouseDeltaY
    }

    override fun getFuckedPredictedPitch(): Float {
        return this.fuckedPredictedPitch
    }

    override fun getFuckedPredictedYaw(): Float {
        return this.fuckedPredictedYaw
    }

    override fun getLastFuckedPredictedPitch(): Float {
        return this.lastFuckedPredictedPitch
    }

    override fun getLastFuckedPredictedYaw(): Float {
        return this.lastFuckedPredictedYaw
    }

    override fun isCinematic(): Boolean {
        return this.cinematic
    }

    override fun getFinalSensitivity(): Double {
        return this.finalSensitivity
    }

    override fun getMcpSensitivity(): Double {
        return this.mcpSensitivity
    }

    override fun getSensitivitySamples(): ArrayDeque<Int?> {
        return this.sensitivitySamples
    }

    override fun getSensitivity(): Int {
        return this.sensitivity
    }

    override fun getMouseDeltaX(): Int {
        return this.mouseDeltaX
    }

    override fun getMouseDeltaY(): Int {
        return this.mouseDeltaY
    }

    override fun getSmoothnessYaw(): Float {
        return smoothnessYaw
    }

    override fun getSmoothnessPitch(): Float {
        return smoothnessPitch
    }

    override fun getConsistency(): Float {
        return consistency
    }

    override fun setCinematic(cinematic: Boolean) {
        this.cinematic = cinematic
    }
}